imgserve.go

```diff
@@ -27,12 +30,60 @@ func main() {
 
 //initRoutes initializes all routes
 func initRoutes(directory string) {
-	http.Handle("/", http.FileServer(http.Dir(directory)))
+	http.Handle("/", logInfo(http.FileServer(http.Dir(directory)), directory))
 	http.HandleFunc("/info/", func(w http.ResponseWriter, r *http.Request) {
 		getImageInfo(w, r, directory)
 	})
 }
 
+func logInfo(next http.Handler, directory string) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		timeStart := time.Now()
+		clientID := "clientID: " + r.RemoteAddr
+		log.Println(clientID + ", start: " + timeStart.Format(time.RFC3339))
+		next.ServeHTTP(w, r)
+
+		timeEnd := time.Now()
+		log.Println(clientID + ", end: " + timeEnd.Format(time.RFC3339))
+
+		imageName := strings.TrimPrefix(r.URL.Path, "/")
+		if imageName == "" {
+			w.WriteHeader(http.StatusBadRequest)
+			w.Write([]byte("name can not be empty"))
+			return
+		}
+
+		averageSpeed, err := countAverageSpeed(directory, imageName, timeEnd.Unix()-timeStart.Unix())
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			w.Write([]byte(fmt.Sprintf("error while calculating average speed, %v", err.Error())))
+			return
+		}
+		log.Println(clientID+", average speed: ", averageSpeed/1000, "kB/s")
+	})
+}
+
+//countAverageSpeed counts average speed of downloading image.
+//it counts sizeOfImage(bytes)/durationOfDownload(seconds)
+func countAverageSpeed(directory, imageName string, duration int64) (int64, error) {
+	if _, ok := imagesInfo[imageName]; !ok {
+		err := getInfo(directory, imageName)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	data := make(map[string]interface{})
+	err := json.Unmarshal(imagesInfo[imageName], &data)
+	if err != nil {
+		return 0, err
+	}
+
+	size := int64(data["virtual-size"].(float64))
```

Comment:
we need to check if we transferred the virtual or the disk size (hopefully the latter) and use the right field